src/replication.c

```diff
@@ -126,37 +126,215 @@ static int handle_mgmt_event_fd(replica_t *replica);
+			REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+			    "for replica(%p)\n", replica);
+			return (ret = -1);
+		}
+	}
+	return ret;
+}
+
+static int
+start_rebuild(void *buf, replica_t *replica, uint64_t data_len)
+{
+	
+	int ret = 0;
+	uint64_t num = 1;
+	zvol_io_hdr_t *rmgmtio = NULL;
+	zvol_op_code_t mgmt_opcode = ZVOL_OPCODE_START_REBUILD;
+	mgmt_cmd_t *mgmt_cmd;
+
+	mgmt_cmd = malloc(sizeof (mgmt_cmd_t));
+	BUILD_REPLICA_MGMT_HDR(rmgmtio, mgmt_opcode, data_len);
+
+	mgmt_cmd->io_hdr = rmgmtio;
+	mgmt_cmd->io_bytes = 0;
+	mgmt_cmd->data = buf;
+	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+
+	MTX_LOCK(&replica->r_mtx);
+	TAILQ_INSERT_TAIL(&replica->mgmt_cmd_queue, mgmt_cmd, mgmt_cmd_next);
+	MTX_UNLOCK(&replica->r_mtx);
+
+	if (write(replica->mgmt_eventfd1, &num, sizeof (num)) !=
+	    sizeof (num)) {
+		REPLICA_NOTICELOG("Failed to inform to mgmt_eventfd "
+		    "for replica(%p)\n", replica);
+		return (ret = -1);
+	}
+	REPLICA_LOG("Start_rebuild opcode sent for Replica ip:%s port:%d "
+	    "state:%d\n", replica->ip, replica->port, replica->state);
+	return ret;
+}
+
+/* Rebuild can be of two type:-
+ * 1. Rebuilding a replica from healthy replica
+ * 2. Rebuilding a replica from all other replica(Mesh rebuild)
+ * API does the following tasks:-
+ * 1. find if there is healthy replica available for rebuilding
+ * 2. find downgraded replica for rebuild
+ * 3. Call prepare_for_rebuild to on healthy replica or on all
+ * other replicas to get IP address and port for replicas.
  */
 void
 update_volstate(spec_t *spec)
 {
+	int ret = 0;
 	uint64_t max;
-	replica_t *replica;
+	replica_t *replica = NULL;
+	replica_t *target_replica = NULL;
+	replica_t *healthy_replica = NULL;
+
+	if (spec->rebuild_in_progress == true) {
+		REPLICA_NOTICELOG("Rebuild is already in progress "
+		    "on replica:%s\n", spec->volname);
+		return;
+	}
+
+	if (((spec->healthy_rcount + spec->degraded_rcount >=
+	    spec->consistency_factor) && (spec->healthy_rcount >= 1)) ||
+	    (spec->healthy_rcount  + spec->degraded_rcount >=
+	    MAX(spec->replication_factor - spec->consistency_factor + 1,
+	    spec->consistency_factor))) { 
 
-	if(((spec->healthy_rcount + spec->degraded_rcount >= spec->consistency_factor) &&
-		(spec->healthy_rcount >= 1))||
-		(spec->healthy_rcount  + spec->degraded_rcount 
-			>= MAX(spec->replication_factor - spec->consistency_factor + 1, spec->consistency_factor))) {
-		if (spec->ready == false)
-		{
-			max = 0;
-			TAILQ_FOREACH(replica, &spec->rq, r_next)
-				max = (max < replica->initial_checkpointed_io_seq) ?
-				    replica->initial_checkpointed_io_seq : max;
+		max = 0;
+		TAILQ_FOREACH(replica, &spec->rq, r_next) {
+			/* Find healthy replica */
+			if ((healthy_replica == NULL) &&
+			    (replica->state == ZVOL_STATUS_HEALTHY)) {
+				healthy_replica = replica;
+			}
+
+			/*
+			 * If we know master replica and healthy
+			 * replica then trigger rebuild
+			 */
+			if ((spec->master_replica != NULL) &&
```

Comment:
we may get 'break' here even with target_replica being NULL, and, with a replica in degraded state.

Reply:
Another issue with 'break' can be is: we may not be finding the 'max' to set io_seq.

Reply:
ok.. regarding my above comment on issue in finding 'max' to set io_seq, I got your point that - with healthy replica being there, we don't need to update io_seq. We will be in 'case 3' of scenarios explained in other comment.

Reply:
is master_replica supposed to be NULL here if 'rebuild_in_progress' is false?